### PR TITLE
show error message if literal is thrown - fixes #661

### LINKED ISSUE #661: t.throws shows `undefined undefined undefined` when a string is thrown
## Description

When using `t.throws()` with a promise that rejects with a string instead of an Error object, the assertion fails.

``` js
import test from 'ava';

async function fn(): Promise<any> {
    return Promise.reject('foo');
}

test(async t => {
    t.throws(fn(), 'foo');
});
```

screen shot 2016-03-21 at 11 53 43

When using `Promise.reject(new Error('foo'));`, it works as expected.

It's due to [these](https://github.com/sindresorhus/ava/blob/master/lib/assert.js#L88-L93) lines. I have no idea why it overrides the `err` reference with a function. I looked into `core-assert` and it doesn't seem like it works with passing in a function as the `expected` parameter. But might be missing something here.
## Environment

Node.js v4.2.1
darwin 15.3.0

### PR DESCRIPTION
**Pull request by SamVerschueren**
This PR fixes #661 which prints `Expected an object to be thrown.`. It's the same error message that ESLint shows when trying to throw a literal. Feel free to provide feedback of any kind.

### CODE REVIEW

**Comment by sindresorhus** [comment]
Can you also add what was thrown to the message, the type and value. Might be helpful.

**Comment by SamVerschueren** [comment]
Also added the types to `actual` and `expected`. Not sure if necessary as it isn't being printed in the console?

**Comment by sindresorhus** [comment]
> Not sure if necessary as it isn't being printed in the console?

Not really necessary as they're only useful in an AssertionError.

**Comment by sindresorhus** [comment]
LGTM

**Comment by SamVerschueren** [comment]
> Not really necessary as they're only useful in an AssertionError.

But it's ok to keep it like this?

File: lib/assert.js [hunk 1]
```diff
@@ -94,13 +94,21 @@ x.throws = function (fn, err, msg) {
...
-		assert.throws(function () {
-			try {
-				fn();
-			} catch (error) {
-				result = error;
-				throw error;
+		try {
+			fn();
+		} catch (error) {
+			if (error && typeof error !== 'object') {
```
Comment (SamVerschueren) [inline lib/assert.js]: Not sure if this check is good enough in every case though. Should I use something like [is-error](https://www.npmjs.com/package/is-error) instead? Because now someone can still `throw {foo: 'bar'}`.

Comment (jamestalmage) [inline lib/assert.js]: :+1: for is-error.

**Comment by novemberborn** [comment]
Hi @SamVerschueren. I don't think this PR is the correct solution to #661.

First some background.

`t.throws()` essentially wraps [`assert.throws()`](https://nodejs.org/api/assert.html#assert_assert_throws_block_error_message). That method supports constructor/instance checks, regular expression matches, and validation functions. We support those checks as well as an error message comparison validation.

[These lines](https://github.com/sindresorhus/ava/blob/efa43690069b3e60526f94f0093082f0c031b79f/lib/assert.js#L88-L93) implement that error message comparison validation. `err` here is not the error thrown by `fn()`, it's how that error should be checked.

In this example the error message does not compare (`'foo'.message !== 'foo'):

```js
import test from 'ava'

async function fn() {
    return Promise.reject('foo');
}

test(async t => {
    t.throws(fn(), 'foo');
});
```

The reason you get `undefined undefined undefined` is that we're [assuming](https://github.com/sindresorhus/ava/blob/efa43690069b3e60526f94f0093082f0c031b79f/lib/assert.js#L108) a validation error like that causes an `AssertionError` to be thrown. However Node [throws the actual error](https://github.com/nodejs/node/blob/be68b68d4863f0d389cc46fdf6f1cbcd1b241d0a/lib/assert.js#L330) which is the `'foo'` string.

In other words we need to check the error coming out of `assert.throws()` and call `test()` with the right arguments.

**Comment by SamVerschueren** [comment]
Hi @novemberborn. Thanks for looking into this. Can you explain why it's not a solution for #661? I read your post multiple times but can't really find an answer :). After looking into it again, I might have found a solution that's "cleaner" and might fit your approach more. I will make another PR with the new solution so we can compare and make sure I understand you correctly.

> [These lines](https://github.com/sindresorhus/ava/blob/efa43690069b3e60526f94f0093082f0c031b79f/lib/assert.js#L88-L93) implement that error message comparison validation. err here is not the error thrown by fn(), it's how that error should be checked.

Yes I know, I had it wrong in the original issue. I'm not touching those lines. I thought that was the issue but after diving deeper I noticed it's a custom compare function for `assert.throws`.

**Comment by novemberborn** [comment]
> Can you explain why it's not a solution for #661?

`t.throws()` shouldn't reject literals. It should ensure *something* was thrown. If that something didn't match expectations it should correctly report that. Right now it's the reporting that's broken, not the validation.

**Comment by jamestalmage** [comment]
> t.throws() shouldn't reject literals. It should ensure something was thrown.

I disagree. https://github.com/sindresorhus/ava/issues/468#issue-128347888

**Comment by SamVerschueren** [comment]
I agree with @jamestalmage although you might argue if AVA should enforce this best practice...

**Comment by jamestalmage** [comment]
> you might argue if AVA should enforce this best practice

I think it protects you from careless mistakes.

**Comment by novemberborn** [comment]
@SamVerschueren hope it's OK if I close this for now. I reckon resolving #661 will require some more changes.
